module/dkg/recovery.go

```diff
@@ -0,0 +1,150 @@
+package dkg
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/state/protocol/events"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/utils/logging"
+)
+
+// BeaconKeyRecovery is a specific module that performs automatic recovery of the random beacon private key
+// when exiting the epoch fallback mode.
+// When following the happy path of the protocol, each node that takes part in the DKG eventually obtains a random beacon
+// private key which is stored in storage.DKGState. When the network enters EFM, then eventually the network will be recovered
+// by submitting the recovery epoch which will effectively exit the EFM. When recovering the next epoch,
+// the operator will use a result of last successful DKG. However, each node needs to track this event and correctly fill
+// the storage.DKGState using the last obtained 'my beacon key' for recovered(next) epoch.
+// This module checks if there is a valid 'my beacon key' for the next epoch, and if not, it tries to recover it from the previous epoch.
+type BeaconKeyRecovery struct {
+	events.Noop
+	log           zerolog.Logger
+	local         module.Local
+	state         protocol.State
+	localDKGState storage.EpochRecoveryMyBeaconKey
+}
+
+var _ protocol.Consumer = &BeaconKeyRecovery{}
+
+// NewBeaconKeyRecovery creates a new BeaconKeyRecovery instance and tries to recover the random beacon private key.
+// This method ensures that we try to recover the random beacon private key even if we have missed the `EpochFallbackModeExited`
+// protocol event (this could happen if the node has crashed between emitting and delivering the event).
+// No errors are expected during normal operations.
+func NewBeaconKeyRecovery(
+	log zerolog.Logger,
+	local module.Local,
+	state protocol.State,
+	localDKGState storage.EpochRecoveryMyBeaconKey,
+) (*BeaconKeyRecovery, error) {
+	recovery := &BeaconKeyRecovery{
+		Noop:          events.Noop{},
+		log:           log.With().Str("module", "my_beacon_key_recovery").Logger(),
+		local:         local,
+		state:         state,
+		localDKGState: localDKGState,
+	}
+
+	err := recovery.tryRecoverMyBeaconPrivateKey(state.Final())
+	if err != nil {
+		return nil, fmt.Errorf("could not recover my beacon private key when initializing: %w", err)
+	}
+
+	return recovery, nil
+}
+
+// EpochFallbackModeExited implements handler from protocol.Consumer to perform recovery of the beacon private key when
+// this node has exited the epoch fallback mode.
+func (b *BeaconKeyRecovery) EpochFallbackModeExited(epochCounter uint64, _ *flow.Header) {
```

Comment:
> reasoning about correctness much simpler (no blocks can be finalized between the EFM exited event being emitted, and this component completing all its work)

I think that is a benefit short-term and tech-debt long-term: 
* Agree that is saves code and complexity now. However, now we are also fully immersed in the topic and arguably best equipped to pursue a more general implementation. 
* In the long term, we have made the following experience repeatedly:
   * We shortcut the more general implementation and correctness argument in the short term. This results in an edge-case that needs to be considered during future work. 
   * When we work on the framework in the future, we have forgotten about the edge case and accidentally break it. 
  
⚠️ Possible scenario here:
* In my mind, the notification design should _only_ guarantees causality (causality here means that the notifications are emitted strictly after the database operations and  order-preserving delivery of events to consumers). It does _not_ guarantee synchronicity within the consensus EventLoop. 
* The current implementation emits events synchronously within the consensus EventLoop. But that is an implementation detail in my opinion. We highlight it as an explanation why processing should be asynchronous, but in my opinion, synchronous event delivery should _not_ be a guaranteed property of the design for the following reason:
   * It is generally possible that some notifications trigger the emissions of other notifications. Unless all processing logic is implemented synchronously (which it isn't for performance reasons) only causality is maintained but not synchronicity.   
* Assume we are finding in the future that synchronously event-emission is notably slowing down consensus. According to the framework specification, it would be totally fine to spin out the event emission to a separate go routine (as it maintains causality). So in this particular case, we might make change the implementation such that the Event Loop channels the events to a separate go routine (after the db operations concluded), and this separate go routine serves the consumer while the EventLoop itself continues processing blocks. This fully preserves causality, but not synchronicity ❗  
  * We have learned in the past that it is a huge lift to go through all implementations and check for assumptions that we potentially could break. In practise we don't do that. 

Long story short:
* I am happy with using the caller's thread to process the notification (this only slows down the HotStuff EventLoop, which is fine given that this code path is used so rarely).
* I would prefer that we do _not_ assume that the source always delivers the notification synchronously on the EventLoop. This would limit massively the flexibility of the framework and generally does not hold in general (`protocol.Consumer` being an exception).

Reply:
### ⚠️ Technically, this consumer needs to be prioritized above other consumers

I think that `EpochFallbackModeExited` conceptually implies that the next epoch is ready. But `EpochFallbackModeExited` could be consumed by multiple consumers and we do not guarantee any order. Furthermore, we have a no clear ordering-convention with `EpochCommittedPhaseStarted`. 

It worried me, because I feel we are writing important state information in response to notifications or asynchronously, which should conceptually already be persisted by the time the notifications is emitted.

I think this is important to get right, because bugs / broken assumptions will be super hard to reproduce and debug.